README.md
[![Build Status](https://travis-ci.org/jrheard/cryptopals.svg?branch=master)](https://travis-ci.org/jrheard/cryptopals)

My implementations of the [CryptoPals](http://cryptopals.com/) challenges. See the `test` directory for proof of completion for each of the challenges I've made it through (not many!).


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-12-23
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-12-23
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/cryptopals/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/cryptopals/compare/0.1.0...0.1.1


test/cryptopals/core_test.clj
(ns cryptopals.core-test
  (:require [clojure.java.io :as io]
            [clojure.string :refer [split]]
            [clojure.test :refer :all]
            [cryptopals.core :refer :all])
  (:import java.util.Base64))

(def FUNKY-MUSIC "I'm back and I'm ringin' the bell \nA rockin' on the mike while the fly girls yell \nIn ecstasy in the back of me \nWell that's my DJ Deshay cuttin' all them Z's \nHittin' hard and the girlies goin' crazy \nVanilla's on the mike, man I'm not lazy. \n\nI'm lettin' my drug kick in \nIt controls my mouth and I begin \nTo just let it flow, let my concepts go \nMy posse's to the side yellin', Go Vanilla Go! \n\nSmooth 'cause that's the way I will be \nAnd if you don't give a damn, then \nWhy you starin' at me \nSo get off 'cause I control the stage \nThere's no dissin' allowed \nI'm in my own phase \nThe girlies sa y they love me and that is ok \nAnd I can dance better than any kid n' play \n\nStage 2 -- Yea the one ya' wanna listen to \nIt's off my head so let the beat play through \nSo I can funk it up and make it sound good \n1-2-3 Yo -- Knock on some wood \nFor good luck, I like my rhymes atrocious \nSupercalafragilisticexpialidocious \nI'm an effect and that you can bet \nI can take a fly girl and make her wet. \n\nI'm like Samson -- Samson to Delilah \nThere's no denyin', You can try to hang \nBut you'll keep tryin' to get my style \nOver and over, practice makes perfect \nBut not if you're a loafer. \n\nYou'll get nowhere, no place, no time, no girls \nSoon -- Oh my God, homebody, you probably eat \nSpaghetti with a spoon! Come on and say it! \n\nVIP. Vanilla Ice yep, yep, I'm comin' hard like a rhino \nIntoxicating so you stagger like a wino \nSo punks stop trying and girl stop cryin' \nVanilla Ice is sellin' and you people are buyin' \n'Cause why the freaks are jockin' like Crazy Glue \nMovin' and groovin' trying to sing along \nAll through the ghetto groovin' this here song \nNow you're amazed by the VIP posse. \n\nSteppin' so hard like a German Nazi \nStartled by the bases hittin' ground \nThere's no trippin' on mine, I'm just gettin' down \nSparkamatic, I'm hangin' tight like a fanatic \nYou trapped me once and I thought that \nYou might have it \nSo step down and lend me your ear \n'89 in my time! You, '90 is my year. \n\nYou're weakenin' fast, YO! and I can tell it \nYour body's gettin' hot, so, so I can smell it \nSo don't be mad and don't be sad \n'Cause the lyrics belong to ICE, You can call me Dad \nYou're pitchin' a fit, so step back and endure \nLet the witch doctor, Ice, do the dance to cure \nSo come up close and don't be square \nYou wanna battle me -- Anytime, anywhere \n\nYou thought that I was weak, Boy, you're dead wrong \nSo come on, everybody and sing this song \n\nSay -- Play that funky music Say, go white boy, go white boy go \nplay that funky music Go white boy, go white boy, go \nLay down and boogie and play that funky music till you die. \n\nPlay that funky music Come on, Come on, let me hear \nPlay that funky music white boy you say it, say it \nPlay that funky music A little louder now \nPlay that funky music, white boy Come on, Come on, Come on \nPlay that funky music \n")

(deftest set-1-challenge-1
  (is (= (hex->base64 "49276d206b696c6c696e6720796f757220627261696e206c696b65206120706f69736f6e6f7573206d757368726f6f6d")
         "SSdtIGtpbGxpbmcgeW91ciBicmFpbiBsaWtlIGEgcG9pc29ub3VzIG11c2hyb29t")))

(deftest set-1-challenge-2
  (is (= (hexify (fixed-xor (unhexify "1c0111001f010100061a024b53535009181c")
                            (unhexify "686974207468652062756c6c277320657965")))
         "746865206b696420646f6e277420706c6179")))

(deftest set-1-challenge-3
  (is (= (second (detect-single-character-xor (unhexify "1b37373331363f78151b7f2b783431333d78397828372d363c78373e783a393b3736")))
         "Cooking MC's like a pound of bacon")))

(deftest set-1-challenge-4
  (is (= (let [inputs (split (slurp (io/resource "set_1_challenge_4.txt")) #"\n")]
           (first (filter identity (map #(second (detect-single-character-xor (unhexify %)))
                                        inputs))))
         "Now that the party is jumping\n")))

(deftest set-1-challenge-5
  (is (= (repeating-key-xor-encrypt (.getBytes "Burning 'em, if you ain't quick and nimble\nI go crazy when I hear a cymbal")
                                    (.getBytes "ICE"))
         (unhexify "0b3637272a2b2e63622c2e69692a23693a2a3c6324202d623d63343c2a26226324272765272a282b2f20430a652e2c652a3124333a653e2b2027630c692b20283165286326302e27282f"))))

(deftest set-1-challenge-6
  (let [input (parse-base64-file "set_1_challenge_6.txt")
        [key-bytes plaintext-bytes] (repeating-key-xor-decrypt input)

        key (apply str key-bytes)
        plaintext (apply str (map char plaintext-bytes))]

    (is (= key "Terminator X: Bring the noise"))

    (is (= plaintext FUNKY-MUSIC))))

(deftest set-1-challenge-7
  (let [input (parse-base64-file "set_1_challenge_7.txt")]
    (is (= (subs (bytes->str (aes-ecb-decrypt input (.getBytes "YELLOW SUBMARINE"))) 0 33)
           "I'm back and I'm ringin' the bell"))))

(deftest set-1-challenge-8
  (let [ciphertexts (as-> "set_1_challenge_8.txt" $
                          (io/resource $)
                          (slurp $)
                          (split $ #"\n")
                          (map #(.decode (Base64/getDecoder) %) $))

        ciphertexts-with-dupes (for [bytes ciphertexts
                                     :when (ciphertext-likely-encrypted-with-ecb-mode? bytes)]
                                 (.encodeToString (Base64/getEncoder) bytes))]

    (is (= ciphertexts-with-dupes
           ["d880619740a8a19b7840a8a31c810a3d08649af70dc06f4fd5d2d69c744cd283e2dd052f6b641dbf9d11b0348542bb5708649af70dc06f4fd5d2d69c744cd2839475c9dfdbc1d46597949d9c7e82bf5a08649af70dc06f4fd5d2d69c744cd28397a93eab8d6aecd566489154789a6b0308649af70dc06f4fd5d2d69c744cd283d403180c98c8f6db1f2a3f9c4040deb0ab51b29933f2c123c58386b06fba186a"]))))

(deftest set-2-challenge-9
  (is (= (pkcs7-pad (.getBytes "YELLOW SUBMARINE") 20)
         '(89 69 76 76 79 87 32 83 85 66 77 65 82 73 78 69 4 4 4 4))))

(deftest set-2-challenge-10
  (let [input (parse-base64-file "set_2_challenge_10.txt")]
    (is (= (bytes->str (aes-cbc-decrypt input
                                        (.getBytes "YELLOW SUBMARINE")
                                        (byte-array (repeat 16 0))))
           FUNKY-MUSIC))))

(deftest set-2-challenge-11
  (let [ecb aes-ecb-encrypt
        cbc aes-cbc-encrypt
        mode-picked (atom nil)]

    (with-redefs [aes-ecb-encrypt #(do
                                     (when (nil? @mode-picked)
                                       (reset! mode-picked :ecb))
                                     (ecb %1 %2))
                  aes-cbc-encrypt #(do
                                     (reset! mode-picked :cbc)
                                     (cbc %1 %2 %3))]

      (let [ciphertext (aes-encrypt-with-random-key-and-padding (.getBytes (apply str (repeat 300 "A"))))]
        (is (= (condp = @mode-picked
                 :ecb true
                 :cbc false)
               (ciphertext-likely-encrypted-with-ecb-mode? ciphertext)))))))

(deftest set-2-challenge-12
  (let [key (generate-aes-key)
        bytes-to-append (base64->bytes "Um9sbGluJyBpbiBteSA1LjAKV2l0aCBteSByYWctdG9wIGRvd24gc28gbXkgaGFpciBjYW4gYmxvdwpUaGUgZ2lybGllcyBvbiBzdGFuZGJ5IHdhdmluZyBqdXN0IHRvIHNheSBoaQpEaWQgeW91IHN0b3A/IE5vLCBJIGp1c3QgZHJvdmUgYnkK")

        encrypt-fn #(aes-ecb-encrypt (pkcs7-pad (concat % bytes-to-append)
                                                16)
                                     key)]
    (is (= (discover-cipher-block-size encrypt-fn) 16))
    (is (true? (does-cipher-use-ecb-mode? encrypt-fn)))

    (is (= (bytes->str (byte-at-a-time-ecb-decrypt encrypt-fn))
           "Rollin' in my 5.0\nWith my rag-top down so my hair can blow\nThe girlies on standby waving just to say hi\nDid you stop? No, I just drove by\n"))))

(deftest set-2-challenge-13
  ; "Using only the user input to profile_for() (as an oracle to generate "valid" ciphertexts)
  ; and the ciphertexts themselves, make a role=admin profile."
  (let [key (generate-aes-key)
        encrypt-fn #(as-> % $
                          (map char $)
                          (apply str $)
                          (profile-for $)
                          (encode-profile $)
                          (.getBytes $)
                          (pkcs7-pad $ 16)
                          (aes-ecb-encrypt $ key))

        ; The ciphertexts we'll be generating and snipping together:
        ;  block           1               2               3
        ;  01234567890123456789012345678901234567890123456789
        ;  email=AAAAAAAAAAAAA&uid=10&role=admin
        ;  email=AAAAAAAAAAAAAAAAAAAAAAAAAAadmin[PADDING]

        ; Ciphertext 1 is used for blocks 1 and 2
        ciphertext-1 (encrypt-fn (apply str (repeat 13 "A")))

        ; Ciphertext 2 gives us block 3, which has "admin" followed by pkcs-7 padding bytes.
        ciphertext-2 (encrypt-fn (str (apply str (repeat 26 "A"))
                                      "admin"
                                      (apply str (repeat 11 (char 11)))))

        payload (concat (take 32 ciphertext-1)
                        (take 16 (drop 32 ciphertext-2)))]

    (is (= (-> payload
               (aes-ecb-decrypt key)
               pkcs7-depad
               bytes->str
               decode-profile)
           {"email" "AAAAAAAAAAAAA" "uid" "10" "role" "admin"}))))


src/cryptopals/core.clj
(ns cryptopals.core
  (:require [clojure.spec.alpha :as s]
            [clojure.string :refer [trim lower-case split]]
            [clojure.java.io :as io]
            [com.rpl.specter :refer [select transform ALL MAP-VALS MAP-KEYS FIRST]]
            [ring.util.codec :as codec])
  (:import java.util.Base64
           (javax.crypto Cipher)
           (javax.crypto.spec SecretKeySpec)))

;;;;;;;;;;;;
;; Utilities

(defn duplicates [xs]
  (for [[k v] (frequencies xs)
        :when (> v 1)]
    [k v]))

; from https://stackoverflow.com/questions/10062967/clojures-equivalent-to-pythons-encodehex-and-decodehex
(defn unhexify [s]
  (map
    (fn [[x y]] (Integer/parseInt (str x y) 16))
    (partition 2 s)))

; same
(defn hexify [bytes]
  (format "%x" (new BigInteger (byte-array bytes))))

; "Cryptopals Rule:
; Always operate on raw bytes, never on encoded strings. Only use hex and base64 for pretty-printing."

(defn hex->base64
  [hex]
  (->> hex
       unhexify
       byte-array
       (.encodeToString (Base64/getEncoder))))

(defn base64->bytes
  [base64]
  (.decode (Base64/getDecoder) base64))

(defn bytes->str [bytes]
  (apply str (map char bytes)))

(defn parse-base64-file
  [filename]
  (as-> filename $
        (io/resource $)
        (slurp $)
        (clojure.string/replace $ #"\n" "")
        (base64->bytes $)))

(defn fixed-xor
  [xs ys]
  (assert (= (count xs) (count ys)))
  (map bit-xor xs ys))

; http://www.macfreek.nl/memory/Letter_Distribution
(def ENGLISH-CHARACTER-FREQUENCIES
  {\a     0.06532
   \b     0.016
   \c     0.0316
   \d     0.0328
   \e     0.1027
   \f     0.0218
   \g     0.0209
   \h     0.0496
   \i     0.0566
   \j     0.0022
   \k     0.0081
   \l     0.0331
   \m     0.0253
   \n     0.0571
   \o     0.06159
   \p     0.0207
   \q     0.001
   \r     0.0498
   \s     0.0531
   \t     0.0751
   \u     0.0268
   \v     0.0106
   \w     0.0183
   \x     0.0019
   \y     0.0172
   \z     0.0011
   \space 0.18})

(defn chi-squared-score
  [expected observed]
  (apply +
         (for [[character expected-occurrences] expected]
           (/ (Math/pow (- expected-occurrences
                           (observed character 0))
                        2)
              expected-occurrences))))

(defn score-string
  "Returns a number, 0 or greater, indicating how likely it is that `s` is some English text. Lower is better."
  [s]
  (let [expected (transform [MAP-VALS] #(* % (count s)) ENGLISH-CHARACTER-FREQUENCIES)
        observed (transform [MAP-KEYS] #(first (seq (lower-case %))) (frequencies s))
        okay-symbols #"[!? .,;\:\-'\"\(\)\n]"
        unrecognized (filter #(and
                                (not (Character/isLetterOrDigit %))
                                (not (re-seq okay-symbols (str %))))
                             s)
        awful (filter #(not (or (= (int %) 10)
                                (<= 32 (int %) 122)))
                      s)]

    (/ (+ (chi-squared-score expected observed)
          (if (seq awful) 10000 0)
          (/ (* (count unrecognized)
                (count unrecognized)
                500)
             (count s)))
       (count s))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Nonstandard encryption/decryption functions from set 1

(defn attempt-to-decode-single-xored-bytes
  [bytes character]
  (let [character-buffer (repeat (count bytes) character)
        xored-bytes (fixed-xor bytes character-buffer)]

    [(char character)
     (bytes->str xored-bytes)
     (score-string (bytes->str xored-bytes))]))

(s/fdef attempt-to-decode-single-xored-bytes
  :ret (s/? (s/tuple char? string? number?)))

(def MAGIC-SINGLE-XOR-DETECTION-STRING-SCORE-THRESHOLD 10)

(defn detect-single-character-xor
  [byte-arr]
  (let [decode-attempts (->> (range 32 123)
                             (map #(attempt-to-decode-single-xored-bytes byte-arr %)))
        valid-attempts (filter #(< (nth % 2) MAGIC-SINGLE-XOR-DETECTION-STRING-SCORE-THRESHOLD)
                               decode-attempts)
        highest-scorer (first (sort-by #(nth % 2)
                                       valid-attempts))]

    (when highest-scorer
      (vec (take 2 highest-scorer)))))

(defn repeating-key-xor-encrypt
  [plaintext-bytes key-bytes]
  (map bit-xor plaintext-bytes (cycle key-bytes)))

(defn hamming-distance
  [a b]
  (assert (= (count a) (count b)))

  ; "The Hamming distance is just the number of differing bits."
  (count (filter #(= % \1)
                 (mapcat #(Integer/toString % 2)
                         (map bit-xor a b)))))

(defn detect-repeating-xor-keysize
  [ciphertext-bytes]
  ; "Let KEYSIZE be the guessed length of the key; try values from 2 to (say) 40."
  (let [key-sizes-and-distances (for [key-size (range 2 41)]
                                  ; "For each KEYSIZE, take the first KEYSIZE worth of bytes, and
                                  ; the second KEYSIZE worth of bytes, and find the edit distance
                                  ; between them."
                                  ;
                                  ; I found that this advice didn't give workable results, so I take the average
                                  ; edit distance between _all_ of the chunks in the ciphertext.
                                  (let [chunks (partition key-size ciphertext-bytes)
                                        distances (map hamming-distance chunks (rest chunks))
                                        mean-distance (/ (apply + distances) (count distances))]

                                    [key-size
                                     ; "Normalize this result by dividing by KEYSIZE."
                                     (float (/ mean-distance key-size))]))]

    ; "The KEYSIZE with the smallest normalized edit distance is probably the key."
    (ffirst (sort-by second key-sizes-and-distances))))

(defn repeating-key-xor-decrypt
  ([ciphertext-bytes]
   (repeating-key-xor-decrypt ciphertext-bytes (detect-repeating-xor-keysize ciphertext-bytes)))

  ([ciphertext-bytes key-size]
    ; "Now that you probably know the KEYSIZE: break the ciphertext into blocks of KEYSIZE length."
   (let [chunked-ciphertext (map (partial apply vector)
                                 (partition key-size ciphertext-bytes))

         ; "Now transpose the blocks: make a block that is the first byte of every block,
         ; and a block that is the second byte of every block, and so on."
         transposed-blocks (for [i (range key-size)]
                             (select [ALL i] chunked-ciphertext))

         ; "Solve each block as if it was single-character XOR. You already have code to do this."
         decodings (map detect-single-character-xor transposed-blocks)

         key (map first decodings)]

     ; "For each block, the single-byte XOR key that produces the best looking histogram
     ; is the repeating-key XOR key byte for that block. Put them together and you have the key."
     [key
      (map bit-xor
           ciphertext-bytes
           (take (count ciphertext-bytes)
                 (cycle (map int key))))])))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; AES encryption/decryption

(defn aes-ecb [mode bytes key-bytes]
  (let [key (SecretKeySpec. key-bytes "AES")
        cipher (Cipher/getInstance "AES/ECB/NoPadding")]

    (.init cipher mode key)
    (->> bytes
         byte-array
         (.doFinal cipher))))

(def aes-ecb-decrypt (partial aes-ecb (Cipher/DECRYPT_MODE)))
(def aes-ecb-encrypt (partial aes-ecb (Cipher/ENCRYPT_MODE)))

(defn pkcs7-pad
  [bytes block-length]
  (let [num-bytes-to-pad (- block-length
                            (rem (count bytes) block-length))]
    (concat bytes
            (repeat num-bytes-to-pad num-bytes-to-pad))))

(defn pkcs7-depad
  [bytes]
  (drop-last (last bytes) bytes))

(defn aes-cbc-encrypt
  ; "In CBC mode, each ciphertext block is added to the next plaintext block before the
  ; next call to the cipher core."
  [plaintext-bytes key-bytes iv-bytes]
  (apply concat
         (reduce (fn [ciphertext-bytes block]
                   (let [xor-block (if (seq ciphertext-bytes)
                                     (last ciphertext-bytes)
                                     iv-bytes)]
                     (conj ciphertext-bytes
                           ; Per https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#Cipher_Block_Chaining_(CBC) :
                           ; When encrypting, xor on the way in and then encrypt.
                           (aes-ecb-encrypt (fixed-xor block xor-block)
                                            key-bytes))))
                 []
                 (partition 16 (pkcs7-pad plaintext-bytes 16)))))

(defn aes-cbc-decrypt
  [ciphertext-bytes key-bytes iv-bytes]
  (let [partitioned (partition 16 ciphertext-bytes)
        xor-blocks (concat [iv-bytes] (drop-last 1 partitioned))
        decryption (apply concat
                          (reduce (fn [plaintext-bytes [block xor-block]]
                                    (conj plaintext-bytes
                                          ; Per https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#Cipher_Block_Chaining_(CBC) :
                                          ; If you're decrypting, decrypt the block and _then_ xor it.
                                          (fixed-xor (aes-ecb-decrypt block key-bytes)
                                                     xor-block)))
                                  []
                                  (map vector partitioned xor-blocks)))]
    (pkcs7-depad decryption)))

(defn generate-aes-key []
  ; "Write a function to generate a random AES key; that's just 16 random bytes."
  (byte-array (take 16 (repeatedly #(rand-int 256)))))

; "Write a function that encrypts data under an unknown key --- that is,
; a function that generates a random key and encrypts under it."
(defn aes-encrypt-with-random-key-and-padding
  [bytes]
  (let [key (generate-aes-key)
        num-chars-to-prepend (+ 5 (rand-int 6))
        num-chars-to-append (+ 5 (rand-int 6))
        padded-bytes (concat (take num-chars-to-prepend (repeatedly #(rand-int 256)))
                             bytes
                             (take num-chars-to-append (repeatedly #(rand-int 256))))]
    (if (= 0 (rand-int 2))
      (aes-ecb-encrypt (pkcs7-pad padded-bytes 16) key)
      (aes-cbc-encrypt padded-bytes key (generate-aes-key)))))

(defn ciphertext-likely-encrypted-with-ecb-mode?
  [ciphertext-bytes]
  (let [chunks (partition 16 ciphertext-bytes)
        dupe-chunks (duplicates chunks)]
    (> (count dupe-chunks) 0)))

(defn discover-cipher-block-size
  [cipher-encrypt-fn]
  ; "Feed identical bytes of your-string to the function 1 at a time --- start with 1 byte ("A"),
  ; then "AA", then "AAA" and so on. Discover the block size of the cipher."
  (loop [i 1
         last-seen-ciphertext-size nil]
    (let [ciphertext (cipher-encrypt-fn (.getBytes (apply str (repeat i "A"))))]

      (if (and last-seen-ciphertext-size
               (not= (count ciphertext) last-seen-ciphertext-size))

        (- (count ciphertext) last-seen-ciphertext-size)
        (recur (inc i) (count ciphertext))))))

(defn does-cipher-use-ecb-mode?
  [cipher-encrypt-fn]
  ; cipher-encrypt-fn must be a one-argument function that takes ciphertext-bytes
  (ciphertext-likely-encrypted-with-ecb-mode?
    (cipher-encrypt-fn (.getBytes (apply str (repeat 300 "A"))))))

(defn byte-at-a-time-ecb-decrypt
  [encrypt-fn]
  (let [cipher-block-size (discover-cipher-block-size encrypt-fn)]
    (loop [decoded-bytes-from-previous-blocks []
           decoded-bytes-from-this-block []
           block-num 0
           ; "Knowing the block size, craft an input block that is exactly 1 byte short
           ; (for instance, if the block size is 8 bytes, make "AAAAAAA"). Think about
           ; what the oracle function is going to put in that last byte position."
           too-short-input (vec (repeat (dec cipher-block-size) (int \A)))]

      (let [ciphertext (take cipher-block-size
                             (drop (* block-num cipher-block-size)
                                   (map int
                                        (encrypt-fn (byte-array too-short-input)))))

            ; "Make a dictionary of every possible last byte by feeding different strings to the oracle;
            ; for instance, "AAAAAAAA", "AAAAAAAB", "AAAAAAAC"."
            encryptions-map (into {}
                                  (for [i (conj (range 122) 10)]
                                    [(->> (concat too-short-input
                                                  decoded-bytes-from-previous-blocks
                                                  decoded-bytes-from-this-block
                                                  [i])
                                          byte-array
                                          encrypt-fn
                                          (map int)
                                          (drop (* block-num cipher-block-size))
                                          (take cipher-block-size))

                                     i]))

            ; "Match the output of the one-byte-short input to one of the entries in your dictionary."
            decoded-byte (encryptions-map ciphertext)]

        (if decoded-byte
          (if (= (count too-short-input) 0)
            ; We've reached the end of a block; add this block's bytes to
            ; decoded-bytes-from-previous-blocks and recur.
            (recur (concat decoded-bytes-from-previous-blocks decoded-bytes-from-this-block [decoded-byte])
                   []
                   (inc block-num)
                   (vec (repeat (dec cipher-block-size) (int \A))))

            ; Otherwise, record this decoded byte in decoded-bytes-from-this-block and recur.
            (recur decoded-bytes-from-previous-blocks
                   (conj decoded-bytes-from-this-block decoded-byte)
                   block-num
                   (vec (rest too-short-input))))

          ; If we couldn't decrypt this byte, we've reached a pkcs-7 padding byte and we're done!
          (concat decoded-bytes-from-previous-blocks decoded-bytes-from-this-block))))))

(defn decode-profile
  [profile-string]
  (codec/form-decode profile-string))

(defn encode-profile
  ; "Write a k=v parsing routine, as if for a structured cookie."
  [profile-map]
  (clojure.string/join "&"
                       (for [[k v] profile-map]
                         (str k "=" v))))

(defn profile-for
  ; "Now write a function that encodes a user profile in that format, given an email address."
  [email]
  ; "Your "profile_for" function should not allow encoding metacharacters (& and =).
  ; Eat them, quote them, whatever you want to do."
  (let [stripped-email (clojure.string/replace email #"[=&]" "")]
    {"email" stripped-email
     "uid"   10
     "role"  "user"}))

(comment

  )

